pkg/fleets/controller_test.go

```diff
@@ -997,9 +1003,129 @@ func TestControllerRollingUpdateDeploymentGSSUpdateFailedErrExpected(t *testing.
+
+	type expected struct {
+		inactiveSpecReplicas int32
+		replicas             int32
+		updated              bool
+	}
+
+	fixtures := map[string]struct {
+		activeStatusReadyReplicas   int32
+		inactiveStatusReadyReplicas int32
+		allocatedReplicas           int32
+		expected                    expected
+	}{
+		"not enough Ready GameServers - do not scale down rest GameServerSet": {
+			activeStatusReadyReplicas:   10,
+			inactiveStatusReadyReplicas: 10,
+			expected: expected{
+				updated:              false,
+				inactiveSpecReplicas: 0,
+				replicas:             75,
+			},
+		},
+		"enough Ready GameServers - scale down rest GameServerSet to Allocated": {
+			activeStatusReadyReplicas:   70,
+			inactiveStatusReadyReplicas: 5,
+			allocatedReplicas:           5,
+			expected: expected{
+				updated:              true,
+				inactiveSpecReplicas: 0,
+				replicas:             70,
+			},
+		},
+		"enough Ready GameServers - scale down rest GameServerSet to 0": {
+			activeStatusReadyReplicas:   70,
+			inactiveStatusReadyReplicas: 10,
+			allocatedReplicas:           0,
+			expected: expected{
+				updated:              true,
+				inactiveSpecReplicas: 0,
+				replicas:             75,
+			},
+		},
+		"scale down rest GameServerSet to > 0": {
+			// 75 - 19 = 56 is minimum number of gameservers
+			// scaling 58 - 56 = -2 gameservers
+			// initial 10 - 2 = 8
+			activeStatusReadyReplicas:   50,
+			inactiveStatusReadyReplicas: 8,
+			allocatedReplicas:           0,
+			expected: expected{
+				updated:              true,
+				inactiveSpecReplicas: 9,
+				replicas:             75,
+			},
+		},
+	}
+
+	for k, v := range fixtures {
+		t.Run(k, func(t *testing.T) {
+			c, m := newFakeController()
+
+			f := defaultFixture()
+			f.Spec.Replicas = 75
+			f.Status.ReadyReplicas = v.activeStatusReadyReplicas + v.inactiveStatusReadyReplicas
+
+			active := f.GameServerSet()
+			active.ObjectMeta.Name = "active"
+			active.Spec.Replicas = 75
+			active.Status.Replicas = 75
+			active.Status.ReadyReplicas = v.activeStatusReadyReplicas
+
+			inactive := f.GameServerSet()
+			inactive.ObjectMeta.Name = "inactive"
+			inactive.Spec.Replicas = 10
+			inactive.Status.Replicas = 10
+			inactive.Status.ReadyReplicas = v.inactiveStatusReadyReplicas
+			inactive.Status.AllocatedReplicas = v.allocatedReplicas
+			updated := false
+			// triggered inside rollingUpdateRest
+			m.AgonesClient.AddReactor("update", "gameserversets", func(action k8stesting.Action) (bool, runtime.Object, error) {
+				updated = true
+				ua := action.(k8stesting.UpdateAction)
+				gsSet := ua.GetObject().(*agonesv1.GameServerSet)
+				assert.Equal(t, inactive.ObjectMeta.Name, gsSet.ObjectMeta.Name)
+				assert.Equal(t, v.expected.inactiveSpecReplicas, gsSet.Spec.Replicas)
+
+				return true, gsSet, nil
+			})
+
+			replicas, err := c.rollingUpdateDeployment(f, active, []*agonesv1.GameServerSet{inactive})
+			assert.Nil(t, err, "no error")
```

Comment:
you can use `require.NoError` here because there is no point to go further if err is returned